tsfc/driver.py

```diff
@@ -357,7 +359,26 @@ def compile_expression_dual_evaluation(expression, to_element, *,
                       index_cache={},
                       scalar_type=parameters["scalar_type"])
 
-    if all(isinstance(dual, PointEvaluation) for dual in to_element.dual_basis()):
+    # A FInAT QuadratureElement with a gem.Variable point set expression is the
+    # target element on the reference cell for dual evaluation where the points
+    # are specified at runtime. This special casing will not be necessary when
+    # FInAT dual evaluation is done - the dual evaluation method of such a
+    # FInAT QuadratureElement will create the necessary gem code.
+    from finat.tensorfiniteelement import TensorFiniteElement
+    runtime_quadrature_rule = (
+        isinstance(finat_to_element, QuadratureElement) or
+        (
+            isinstance(finat_to_element, TensorFiniteElement) and
+            isinstance(finat_to_element.base_element, QuadratureElement)
+        ) and
+        isinstance(finat_to_element._rule.point_set.expression, gem.Variable)
+    )
+
+    point_evaluation_nodes_only = (
+        all(isinstance(dual, PointEvaluation) for dual in to_element.dual_basis()) and not
+        runtime_quadrature_rule
+    )
+    if point_evaluation_nodes_only:
```

Comment:
And this is a similar, but pre-existing anti-pattern.  I wish to see this gone, too.

Reply:
See previous comment - I'm not going to refactor what is going to be replaced soon anyway and is beyond the scope of this PR.